### PR TITLE
Add new Hack file extensions.

### PR DESCRIPTION
**Pull request by azjezz**
## Description

Add support for hack `hhi` extension.

## Checklist:
- [x] **I am associating a language with a new file extension.**
  - [x] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      - `.hhi` : https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Ahhi+hh
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [Map.hhi](https://github.com/facebook/hhvm/blob/3df035714e281ea56d2d99592757277fb8e44401/hphp/hack/src/utils/findUtils.ml#L16)

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -1904,7 +1904,10 @@ Hack:
   codemirror_mode: php
   codemirror_mime_type: application/x-httpd-php
   extensions:
+  - ".hack"
+  - ".hck"
```
Comment (pchaigno) [inline lib/linguist/languages.yml]: Usage for that extension is [still too low](https://github.com/search?q=extension%3Ahck+NOT+sdhdfjdhfjd&type=Code). Could you remove it?

**Comment by azjezz** [comment]
@pchaigno i have removed the `.hck` extension, also i didn't add heuristic changes as the `.hack` and `.hhi` extensions are not used by any other language

**Comment by azjezz** [comment]
cc @pchaigno

**Comment by pchaigno** [comment]
Usage for the `.hhi` extension seems fine. I've counted 830 repositories by 746 users using [Harvester](https://github.com/Alhadis/Harvester) and [the `extension:hhi hh` search query](https://github.com/search?o=desc&q=extension%3Ahhi+hh&s=&type=Code).

We'll have an issue with the `.hack` extension however, if we merge this pull request as-is. There are [plenty of `.hack` files that contain just zeros and ones](https://github.com/search?o=desc&q=extension%3Ahack+NOT+hh&s=indexed&type=Code) which will be classified as Hack with this pull request. I don't know what these files are. Either we can identify them and classify them as their own (data?) language, or we'll need to remove the `.hack` extension from Hack for the moment.

**Comment by azjezz** [comment]
@pchaigno 

#### Problems with `.php` extension :

currently its hard to detect hack files that use `.php` extension, for example i noticed that today my repository was reporting `11%` hack, when its a pure php app. the issue was that some `.php` files contained only HTML, since there's no `<?php`, linguist detects them as `hack`.

see : https://github.com/azjezz/web-php-mock-ups

#### Problems with `.hh` extension :
hh files used to be detected as `c++` header files, but since we added a heuristic rule last time to check for `<?hh` it was fixed.

the issue here is that hack is not planning to keep the `<?hh` tag after `hhvm 4.0` 
#### Solution : 
`.hack` seems to be the only reasonable solution.
there's no other language using this extension, makes it easy to detect ( note: there's no `<?hh` tag in `.hack` files )

#### Solution for 01 files
the solution would be to  identify them and classify them as their own (data or binary) language.

---

cc @fredemmott

**Comment by fredemmott** [comment]
We’ve not yet announced support for .hack files - an while the support was committed, it is about to change, and these pull request examples will no longer be considered valid - as of next monday’s release, having any ‘<?’ token in a .hack file is a syntax error. They’re always strict mode.

**Comment by azjezz** [comment]
@fredemmott i suggested these examples before we discussed the `<?hh` tag in `.hack` files :stuck_out_tongue: i have removed the `<?hh` tag.

**Comment by jeffomatic** [comment]
Just for context: the existing `.hack` files are almost overwhelmingly going to be from solutions to the [Nand2Tetris](https://www.nand2tetris.org/) course, a popular MOOC for computer science (I recommend it!). `.hack` is a "compiled" machine code format for the fictitious CPU that students implement. As such, it's not a language that can be syntax-highlighted in any meaningful way. The content of those files are ASCII `1`'s and `0`'s, plus whitespace.

**Comment by pchaigno** [comment]
Thank you all for the information!

I've checked 2817 `.hack` files from [the `extension:hack 0000000000010000 OR 0000000000000000` search query](https://github.com/search?q=extension%3Ahack+0000000000010000+OR+0000000000000000) using Harvester and found 931 repositories among 918 users. It's enough to add support for that second language. Once we've added that language, it shouldn't be hard to distinguish between the two with 100% accuracy :-)
@jeffomatic Any idea what we should name it? Looks like the official name is Hack, so maybe "Nand2Tetris Hack"?

> @fredemmott i suggested these examples before we discussed the `<?hh` tag in `.hack` files  i have removed the `<?hh` tag.

We try to follow usage and not specifications in Linguist. If one language says some syntax is illegal but a large number of users still rely on that syntax, we'll try to support it. So I'm not sure we should remove the tags here. The best thing would be to preserve the sample files in their original form.

**Comment by fredemmott** [comment]
> We try to follow usage and not specifications in Linguist. If one language says some syntax is illegal but a large number of users still rely on that syntax, we'll try to support it. So I'm not sure we should remove the tags here. The best thing would be to preserve the sample files in their original form.

There's definitely not a large number of users; Hack uses a of variant PHP-style autoloading, and no release of the autoloader supports the `.hack` extension *at all* - and it's only been in master [since Jan 15th](https://github.com/hhvm/hhvm-autoload/commits/master). Any usage of `<?hh` is extremely early adopters running a locally patched version or master from the last two weeks - as far as I'm aware, it's just @azjezz.

**Comment by fredemmott** [comment]
Though, `<?hh` should remain supported in `.php`, `.hh`, `.hhi` files etc.

**Comment by azjezz** [comment]
@fredemmott the `.hack` files in @nuxed were generated using `hh-to-hack` so they don't use `<?hh` already

**Comment by jeffomatic** [comment]
> @jeffomatic Any idea what we should name it? Looks like the official name is Hack, so maybe "Nand2Tetris Hack"?

I think that's fine. It is pretty unlikely that anyone would intentionally configure a language context for these files since they are far closer to binaries than they are human-consumable text.

Caveat that I don't speak for the Nand2Tetris project or community in any real way other than having gone through the course as a hobbyist.

**Comment by pchaigno** [comment]
Thanks @jeffomatic!

@azjezz - Could you add the Nand2Tetris Hack language to Linguist in this pull request as well, with a [heuristic](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.yml) to disambiguate between the two `.hack` language? I think we can probably add Nand2Tetris as a data language.

Something along the lines of:
```yaml
- extensions: ['.hack']
  rules:
  - language: Nand2Tetris Hack
    pattern: ^[01]+$
  - language: Hack
```
should be enough.

**Comment by azjezz** [comment]
@pchaigno done :+1:

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -1904,7 +1904,9 @@ Hack:
   codemirror_mode: php
   codemirror_mime_type: application/x-httpd-php
   extensions:
+  - ".hack"
```
Comment (pchaigno) [inline lib/linguist/languages.yml]: Is `.hack` considered the primary extension for Hack now?

Comment (azjezz) [reply lib/linguist/languages.yml]: yes

File: lib/linguist/languages.yml [hunk 3]
```diff
@@ -2986,6 +2988,12 @@ Myghty:
   tm_scope: none
   ace_mode: text
   language_id: 239
+Nand2Tetris Hack:
+  type: data
+  ace_mode: text
```
Comment (pchaigno) [inline lib/linguist/languages.yml]: You'll need to add `tm_scope: none` to tell Linguist we don't have a syntax highlighting grammar for that language.

File: lib/linguist/languages.yml [hunk 4]
```diff
@@ -2986,6 +2988,12 @@ Myghty:
   tm_scope: none
   ace_mode: text
   language_id: 239
+Nand2Tetris Hack:
```
Comment (pchaigno) [inline lib/linguist/languages.yml]: Languages are sorted with uppercase letters first.

File: test/test_language.rb [hunk 5]
```diff
@@ -343,6 +345,8 @@ def test_extensions
     assert Language['LSL'].extensions.include?('.lsl')
     assert Language['Perl'].extensions.include?('.pl')
     assert Language['Python'].extensions.include?('.py')
+    assert Language['Hack'].extensions.include?('.hack')
+    assert Language['Hack'].extensions.include?('.hhi')
```
Comment (pchaigno) [inline test/test_language.rb]: You can remove these tests. We already have tests in place to check this. I'm not even sure what these hardcoded tests are used for.

Comment (azjezz) [reply test/test_language.rb]: :+1:

File: lib/linguist/languages.yml (hunk 4 above)
Comment (azjezz) [reply lib/linguist/languages.yml]: :+1:

File: lib/linguist/languages.yml (hunk 3 above)
Comment (azjezz) [reply lib/linguist/languages.yml]: :+1:

File: samples/Hack/EmptyPipelineException.hack [hunk 6]
```diff
@@ -1,7 +1,7 @@
-namespace Waffle\Http\Server\Exception;
+namespace Nuxed\Http\Server\Exception;
```
Comment (pchaigno) [inline samples/Hack/EmptyPipelineException.hack]: Why the changes for the two Hack sample files?

File: test/test_language.rb [hunk 7]
```diff
@@ -177,6 +177,8 @@ def test_find_by_extension
     assert_equal [Language['Ruby']], Language.find_by_extension('foo/bar.rb')
     assert_equal [Language['Ruby']], Language.find_by_extension('PKGBUILD.rb')
     assert_equal ['C', 'C++', 'Objective-C'], Language.find_by_extension('foo.h').map(&:name).sort
+    assert_equal [Language['Hack']], Language.find_by_extension('HomeController.hack')
+    assert_equal [Language['Hack']], Language.find_by_extension('string_functions.hhi')
```
Comment (pchaigno) [inline test/test_language.rb]: Same thing for these tests.

File: samples/Hack/EmptyPipelineException.hack (hunk 6 above)
Comment (azjezz) [reply samples/Hack/EmptyPipelineException.hack]: the waffle project is deprecated and nuxed uses more hack features, might help in the future when updating the syntax highlight for hack :stuck_out_tongue: ( PHP Syntax highlight is not working well - but its not a priority now )

File: test/test_language.rb (hunk 7 above)
Comment (azjezz) [reply test/test_language.rb]: :+1:

File: samples/Hack/EmptyPipelineException.hack (hunk 6 above)
Comment (pchaigno) [reply samples/Hack/EmptyPipelineException.hack]: I'd rather we did not change the existing sample files. They are not meant to be execute, but simply to represent in-the-wild usage of Hack.

**Comment by pchaigno** [comment]
My guess would be that the tests are failing because the tokenizer is unable to extract tokens from the Nand2Tetris hack sample file. If that's the case, we'll probably need to adapt the tokenizer or add an exception mechanism. I have found time to debug this yet.

**Comment by Lexidor** [comment]
@pchaigno if the Nand2Tetris files are only ones and zeroes (and unlimited whitespace), then there are no real tokens. This _language_ is not read as source code. Would it be okay if I did something dirty and just specified the token definition as "Each byte is the start of a new token"? Would this somehow cause something unrelated to break? I would like to have syntax highlighting in hack files, so sorry for not prioritising the Nand2Tetris hack format more.

**Comment by Lexidor** [comment]
@pchaigno Are we following the route of not requiring tokenizers. Tokenizing such files is rather pointless, since there are no tokens. Hack code is being used in dothack files in many more places than it was at the release of HHVM4. This leaves entire repositories as black text on GitHub.

**Comment by azjezz** [comment]
cc @pchaigno, any updates here ? after #4544 , the `.hack` files syntax is "highlightable" when using a gitattribute override to classify `.hack` as `Hack`.

example : [`.gitattribute`](https://github.com/nuxed/framework/blob/rx/.gitattributes) / [`Translator.hack`](https://github.com/nuxed/framework/blob/rx/src/Nuxed/Translation/Translator.hack)

**Comment by lildude** [comment]
Wow! How did I miss this one for so long. Sorry folks. 

Support for the `.hack` extension has now been added in https://github.com/github/linguist/pull/4588 and we completely forgot about the Nand2Tetris files. I don't think anyone has noticed... or really cares about those files being identified as Hack (famous last words 😆) so lets update this branch and change it to only add support for the `.hhi` extension.

**Comment by azjezz** [comment]
@lildude i myself forgot about this PR :) 

i have update this PR to only address `.hhi` extension.
